src/Telemetry/Microsoft.DotNet.AzureDevOpsTimelineTests/AzureDevOpsTimelineTests.cs

```diff
@@ -0,0 +1,308 @@
+                l.AddProvider(new NUnitLogger());
+            });
+
+            collection.Configure<AzureDevOpsTimeline>(options =>
+            {
+
+            });
+
+            collection.AddSingleton(systemClock);
+            collection.AddSingleton(timelineTelemetryRepository);
+            collection.AddSingleton(azureDevOpsClient);
+            collection.AddScoped<AzureDevOpsTimeline>();
+
+            ServiceProvider services = collection.BuildServiceProvider();
+            return new TestData(services.GetRequiredService<AzureDevOpsTimeline>(), services);
+        }
+
+        /// <summary>
+        /// Test a happy path. Database is empty, so all returned builds will be added.
+        /// </summary>
+        [Test]
+        public async Task VanillaTest()
+        {
+            // Test input
+            Dictionary<Build, List<Timeline>> builds = new Dictionary<Build, List<Timeline>>()
+            {
+                {
+                    new Build()
+                    {
+                        Id = 1,
+                        Project = new TeamProjectReference() { Name = "public" },
+                        ValidationResults = Array.Empty<BuildRequestValidationResult>(),
+                        Reason = "pullRequest",
+                        Parameters = "{\"system.pullRequest.targetBranch\": \"theTargetBranch\"}"
+                    },
+                    new List<Timeline>() {
+                        new Timeline()
+                        {
+                            Id = "1",
+                            LastChangedOn = DateTimeOffset.Now,
+                            Records = new[]
+                            {
+                                new TimelineRecord()
+                                {
+                                    Id = "1",
+                                    Issues = Array.Empty<TimelineIssue>()
+                                },
+                                new TimelineRecord()
+                                {
+                                    Id = "2",
+                                    Issues = Array.Empty<TimelineIssue>()
+                                },
+                            }
+                        }
+                    }
+                }
+            };
+
+            // Test setup
+            MockAzureClient mockAzureClient = new MockAzureClient(builds);
+            InMemoryTimelineTelemetryRepository telemetryRepository = new InMemoryTimelineTelemetryRepository(new List<(string project, DateTimeOffset latestTime)>());
+
+            using TestData testData = SetUp(telemetryRepository, mockAzureClient, new SystemClock());
+
+            /// Test execution
+            await testData.Controller.RunProject("public", 1000, CancellationToken.None);
+
+            // Test results
+            telemetryRepository.TimelineBuilds.Should().HaveCount(1);
+            telemetryRepository.TimelineBuilds[0].Build.Should().BeSameAs(builds.Single().Key);
```

Comment:
I think you can use the ".SatisfyRespectivly" on the collection to do the collection asserting in one swoop.

Reply:
Neat!